backend/src/apiserver/model/job.go

```diff
@@ -14,24 +14,168 @@
+	default:
+		return StatusStateUnspecifiedV1
+	}
+}
+
+// Converts to v2beta1-compatible internal representation of job status.
+// This should be called before converting to v2beta1 API type.
+func (s StatusState) ToV2() StatusState {
+	switch s.toUpper() {
+	case StatusStateUnspecified, StatusStateUnspecifiedV1, "MODE_UNSPECIFIED", "NO_STATUS", "":
+		return StatusStateUnspecified
+	case StatusStateEnabled, "ENABLE", "READY", "RUNNING", "SUCCEEDED":
+		return StatusStateEnabled
+	case StatusStateDisabled, "DISABLE":
+		return StatusStateDisabled
+	default:
+		return StatusStateUnspecified
+	}
+}
 
 type Job struct {
-	UUID               string `gorm:"column:UUID; not null; primary_key"`
-	DisplayName        string `gorm:"column:DisplayName; not null;"` /* The name that user provides. Can contain special characters*/
-	Name               string `gorm:"column:Name; not null;"`        /* The name of the K8s resource. Follow regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?'*/
-	Namespace          string `gorm:"column:Namespace; not null;"`
-	ServiceAccount     string `gorm:"column:ServiceAccount; not null;"`
-	Description        string `gorm:"column:Description; not null"`
-	MaxConcurrency     int64  `gorm:"column:MaxConcurrency;not null"`
-	NoCatchup          bool   `gorm:"column:NoCatchup; not null"`
-	CreatedAtInSec     int64  `gorm:"column:CreatedAtInSec; not null"` /* The time this record is stored in DB*/
-	UpdatedAtInSec     int64  `gorm:"column:UpdatedAtInSec; not null"`
-	Enabled            bool   `gorm:"column:Enabled; not null"`
+	UUID           string `gorm:"column:UUID; not null; primary_key;"`
+	DisplayName    string `gorm:"column:DisplayName; not null;"` /* The name that user provides. Can contain special characters*/
+	K8SName        string `gorm:"column:Name; not null;"`        /* The name of the K8s resource. Follow regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?'*/
+	Namespace      string `gorm:"column:Namespace; not null;"`
+	ServiceAccount string `gorm:"column:ServiceAccount; not null;"`
+	Description    string `gorm:"column:Description; not null;"`
+	MaxConcurrency int64  `gorm:"column:MaxConcurrency; not null;"`
+	NoCatchup      bool   `gorm:"column:NoCatchup; not null;"`
+	CreatedAtInSec int64  `gorm:"column:CreatedAtInSec; not null;"` /* The time this record is stored in DB*/
+	UpdatedAtInSec int64  `gorm:"column:UpdatedAtInSec; not null;"`
+	Enabled        bool   `gorm:"column:Enabled; not null;"`
+	ExperimentId   string `gorm:"column:ExperimentUUID; not null;"`
+	// ResourceReferences are deprecated. Use Namespace, ExperimentId
+	// PipelineSpec.PipelineId, PipelineSpec.PipelineVersionId
 	ResourceReferences []*ResourceReference
 	Trigger
 	PipelineSpec
-	Conditions string `gorm:"column:Conditions; not null"`
+	Conditions string `gorm:"column:Conditions; not null;"`
+}
+
+// Converts to v1beta1-compatible internal representation of job.
+// This should be called before converting to v1beta1 API type.
+func (j *Job) ToV1() *Job {
```

Comment:
I thought I left a comment/question on these `ToV1()` and `ToV2()` methods, couldn't find it now, so not sure if you've seen it.
If I understand correctly, this is how you want to achieve v2 api being able to get v1 entities, and v1 api being able to get v2 entities? 
While I might be wrong, but I think maybe only LIST apis need to do so, while GET apis (and CREATE, UPDATE apis, but I don't think they're related to these conversion method?) don't need this compatible support in practice. There's always difference between v1 and v2 entities, for instance, v1 entities doesn't have IR, while v2 entities doesn't have Argo workflow spec. With that difference remains, having the resource references part converted back and forth may not actually provide much value. So I think maybe we can simply error gracefully when calling v1 GET on v2 entities and v2 GET on v1 entities.

Adding more folks to see if they may have different thoughts here.
/cc @Linchin 
/cc @jlyaoyuli

Reply:
> On a second thought, I don't think this has anything to do with e2e test failure. But still, since both v1 and v2 (no-suffix) exist in this PR, and the interfaces are identical, shouldn't we use v2 now?

`ReportWorkflow` is no implemented in the [master branch](https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/report_server.go#L32). I might be missing something, but all FE integrations tests should be failing because of this. Created runs cannot be updates and should stay in "Running" state forever.

Reply:
> If they are valid values from v1, should they all be defined as constants? having a mix use of constants and values is anti-best practice, unless the code is directly copied/moved from existing code, we should avoid creating such pattern. Secondly, the values look very suspicious. For example, oo we really have "READY" and "RUNNING" state for job?

I agree. Let me see if I can minimize the number of cases here. 

Regarding suspicious values. In v1 API, Job has `Mode`, `Status` and `Enabled`. I guess, `Mode` and `Enabled` are duplicating fields specified by a user, while the `Status` is the output status. It is a STRING instead of ENUM, which is the reason I added these cases.

When a persistent agent reports a `ScheduledWorkflow`, it [updates](https://github.com/kubeflow/pipelines/blob/9a32f765ea60f1e102adefe012640ee89d840b97/backend/src/apiserver/storage/job_store.go#L366) the column "Conditions" with `swf.ConditionSummary()`. This value is later used in v1 API Status [here](https://github.com/kubeflow/pipelines/blob/9a32f765ea60f1e102adefe012640ee89d840b97/backend/src/apiserver/server/api_converter.go#L412). `ConditionSummary()` returns `NO_STATUS` or the `Type` of the latest condition. Values for the conditions are coming from [here](https://github.dev/kubeflow/pipelines/blob/9a32f765ea60f1e102adefe012640ee89d840b97/backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/types.go#L173) that include `Enabled`, `Disabled`, `Running`, `Succeeded`, `Error`. There are multiple unit tests with "Ready" and "Running" as values for Status field. Hence, I included them here.

V2 API conversion function does not take that into consideration [here](https://github.com/kubeflow/pipelines/blob/9a32f765ea60f1e102adefe012640ee89d840b97/backend/src/apiserver/server/api_converter.go#L502). So we may see jobs returned with `STATUS_UNSPECIFIED` in V2 if we don't include these values.

Please, correct me if I am wrong.

Reply:
>I thought I left a comment/question on these ToV1() and ToV2() methods, couldn't find it now, so not sure if you've seen it. If I understand correctly, this is how you want to achieve v2 api being able to get v1 entities, and v1 api being able to get v2 entities? While I might be wrong, but I think maybe only LIST apis need to do so, while GET apis (and CREATE, UPDATE apis, but I don't think they're related to these conversion method?) don't need this compatible support in practice. There's always difference between v1 and v2 entities, for instance, v1 entities doesn't have IR, while v2 entities doesn't have Argo workflow spec. With that difference remains, having the resource references part converted back and forth may not actually provide much value. So I think maybe we can simply error gracefully when calling v1 GET on v2 entities and v2 GET on v1 entities.  Adding more folks to see if they may have different thoughts here. /cc @Linchin /cc @jlyaoyuli

The goals were:
1. Store everything in v2-compatible model. This requires mapping from v1 values to v2. The benefit is ability of users to work with v1 resources after upgrading to v2. IMO, this is strategically important if we want user to migrate. Otherwise, we will end up with users who don't want to upgrade dues to lack of support for their v1 resources (experiments, pipelines, jobs, runs, etc.).
2. Enable support for v1 as much as possible to not break anything. This is also strategically important to enable a smooth transition for the users.

We should not aim to have identical representations of v1 and v2 resource, as there are fundamental differences. The discrepancy between different fields in v1 and v2 that you pointed out make sense. However, we don't have to limit users from getting v1 resource via v2 sdk and vice versa. Besides manifests, there are many other important fields that users may be interested in (e.g. status/state, names, timestamps, parent/child resources, etc.). Returning an error will push away users from KFP v2.  

Your earlier point about limiting the use of `ToV1()` and `ToV2()` makes sense too. I will take a look on how to improve it.